### DISCUSSION TITLE
MQTT Werte schreiben

### DISCUSSION THREAD

**Discussion by ubik128** [open3e/open3e#167] [Q&A]
Hallo!

Bei mir funktioniert schon lesen über MQTT (über loxberry MQTT Gateway und dann http Nachricht)
und schreiben von einem einfachen Wert (zb.: Warmwasser Solltemperatur) mit "396": "DomesticHotWaterTemperatureSetpoint" mit folgender MQTT Nachricht
open3e/cmnd {"mode": "write", "data":[[396,51]]}

Jetzt meine Frage:
Wie kann ich etwas "komplexeres" schicken?
z.B.: Raumtemperatursolltemperatur?
424 MixerOneCircuitRoomTemperatureSetpoint {"Comfort": 22.0, "Standard": 22.0, "Reduced": 21.0,
 "Unknown2": "0000", "Unknown1": 0} 

open3e/cmnd {"mode": "write", "data": [[424, '{"Comfort": 22.0, "Standard": 22.0, "Reduced": 21.0, "Unknown2": "0000", "Unknown1": 0}']]}
--> hat leider nicht funktioniert

Und Schnellwahl Warmwasser würde ich auch gerne auch wissen:
1006 TargetQuickMode "02000000" ....normal Betrieb
1006 TargetQuickMode "02010000" ....Einmal Warmwasser

Danke Christian

**Comment by ubik128** [comment]
OK Frage selber geklärt :)
open3e/cmnd {"mode": "write", "data": [[424, {"Comfort": 22.0, "Standard": 22.0, "Reduced": 21.0, "Unknown2": "0000", "Unknown1": 0}]]}
--> die einfachen Gänsefüsse bei {} waren falsch.
Jetzt scheint es zu funktionieren :)

**Reply by philippoo66** [reply]
moin @ubik128 !

jap, ich brauch auch immer ein paar Anläufe bis ich das fehlerfrei hinbekomme. Wir haben aber grad den ["149er" Branch](https://github.com/open3e/open3e/tree/149-feature-request-writing-to-complex-dids) in der beta-Test Phase. Damit kannst du einzelne "Sub-Items" lesen und schreiben. Nebenbei können die auch noch mit Namen adressiert werden, z.B. sowas 
`{"mode":"write", "addr":"0x680", "data":[["BusIdentification.BusType","CanExternal"]]}` (das haben wir aber mit der [virtualE3](https://github.com/philippoo66/vitualE3) gemacht, bei realer Anlage bitte was nehmen, das auch zum Schreiben gedacht ist).

Wichtig beim Adressieren von Subs ist, dass immer auch die ECU Adresse mit angegeben wird, also `0x680.256.3`, weil wir sonst ECU.DID nicht von DID.Sub unterscheiden können. Zumindest ist das bei der command line so, bei MQTT hab ich das noch garnicht getestet, da könnte es sogar so funktionieren, weil die ECU Addr schon immer extra war, und ohne die Default ECU genommen wird.

Wenn du Lust hast, uns beim Testen zu unterstützen - immer gern!
